generic-ec/src/multiscalar/straus.rs

```diff
@@ -0,0 +1,327 @@
+        }
+        Self(table)
+    }
+    /// Takes odd integer $x$ such as $0 < x < 2^4$, returns $x P$
+    fn get(&self, x: usize) -> Point<E> {
+        debug_assert_eq!(x & 1, 1);
+        debug_assert!(x < 16);
+
+        self.0[x / 2]
+    }
+}
+
+/// Stores a width-$w$ "Non-Adjacent Form" (NAF) of multiple scalars
+///
+/// Width-$w$ NAF represents an integer $k$ via coefficients $k_0, \dots, k_n$ such as:
+///
+/// $$k = \sum_{i=0}^{n} k_i \cdot 2^i$$
+///
+/// where each $k_i$ is odd and lies within range $-2^{w-1} \le k_i < 2^{w-1}$.
+///
+/// Non Adjacent Form allows us to reduce size of tables we need to precompute in Straus
+/// multiscalar multiplication by factor of 4.
+struct NafMatrix<E: Curve> {
+    /// Size of one scalar in non adjacent form
+    naf_size: usize,
+    /// Input parameter `w`
+    w: usize,
+    /// width = 2^w
+    width: u64,
+    /// width_half = width / 2
+    width_half: u64,
+    /// window_mask = width - 1
+    window_mask: u64,
+    matrix: Vec<i8>,
+
+    _curve: core::marker::PhantomData<E>,
+}
+
+impl<E: Curve> NafMatrix<E> {
+    /// Construct a new matrix with parameter `w`
+    ///
+    /// Preallocates memory to fit `capacity` amount of scalars
+    fn new(w: usize, capacity: usize) -> Self {
+        assert!((2..=8).contains(&w));
+        let naf_size = Scalar::<E>::serialized_len() * 8 + 1;
+        let width = 1 << w;
+
+        Self {
+            naf_size,
+            w,
+            width,
+            width_half: 1 << (w - 1),
+            matrix: Vec::with_capacity(naf_size * capacity),
+            window_mask: width - 1,
+            _curve: Default::default(),
+        }
+    }
+    /// Adds a scalar into matrix
+    fn add_scalar(&mut self, scalar: &Scalar<E>) {
+        let scalar_bytes = scalar.to_le_bytes();
+        let mut x_u64 = vec![0u64; scalar_bytes.len() / 8 + 1];
+        read_le_u64_into(&scalar_bytes, &mut x_u64[0..4]);
+
+        let offset = self.matrix.len();
+        debug_assert!(
+            offset + self.naf_size <= self.matrix.capacity(),
+            "unnecessary allocations detected"
+        );
+        self.matrix.resize(offset + self.naf_size, 0i8);
+        let naf = &mut self.matrix[offset..];
+
+        let mut pos = 0;
+        let mut carry = false;
+        while pos < self.naf_size {
+            let u64_idx = pos / 64;
+            let bit_idx = pos % 64;
+            let bit_buf: u64 = if bit_idx < 64 - self.w {
+                // This window bits are contained in a single u64
+                (x_u64[u64_idx] >> bit_idx) & self.window_mask
+            } else {
+                // Combine the current u64's bits with the bits from the next u64
+                ((x_u64[u64_idx] >> bit_idx) | (x_u64[u64_idx + 1] << (64 - bit_idx)))
+                    & self.window_mask
+            };
+
+            // Add the carry into the current window
+            let window = if carry { bit_buf + 1 } else { bit_buf };
```

Comment:
`let window = bit_buf + u64::from(carry);` - though both are ok

Reply:
That's what I wrote at first but then I thought that checking bool could be cheaper than converting it to u64 and adding to the bit_buf

Reply:
I'm pretty sure they both get optimized to the same `test carry carry; repz inc`, it's more a question of style